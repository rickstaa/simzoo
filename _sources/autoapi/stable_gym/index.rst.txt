stable_gym
==========

.. py:module:: stable_gym

.. autoapi-nested-parse::

   Entry_point that register the Stable Gym gymnasium environments.



Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/stable_gym/common/index
   /autoapi/stable_gym/envs/index


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/stable_gym/version/index


Attributes
----------

.. autoapisummary::

   stable_gym.__version__
   stable_gym.__version_tuple__
   stable_gym.ENVS


Classes
-------

.. autoapisummary::

   stable_gym.MaxEpisodeStepsInjectionWrapper


Package Contents
----------------

.. py:class:: MaxEpisodeStepsInjectionWrapper(env)

   Bases: :py:obj:`gymnasium.Wrapper`


   A gymnasium wrapper that injects the ``max_episode_steps`` attribute of the
   ``TimeLimit`` wrapper into the base environment as the
   ``_time_limit_max_episode_steps`` attribute. If the environment is not wrapped in
   a ``TimeLimit`` wrapper, then the ``_time_limit_max_episode_steps`` attribute is
   set to ``None``.

   Wrap a gymnasium environment.
   :param env: The gymnasium environment.
   :type env: gym.Env


   .. py:attribute:: max_episode_steps


.. py:data:: __version__
   :value: '0.20.7'


.. py:data:: __version_tuple__

.. py:data:: ENVS

