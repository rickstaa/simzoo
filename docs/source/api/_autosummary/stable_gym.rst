stable\_gym
===========

.. automodule:: stable_gym
  
   
   
   

   
   
   

   
   
   

   
   
   



.. rubric:: Modules

.. autosummary::
   :toctree:
   :template: custom-module-template.rst
   :recursive:

   stable_gym.common
   stable_gym.envs

