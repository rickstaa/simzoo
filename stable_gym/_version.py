# Stores the package version number so that it can be accessed from other modules.
version = "0.6.4"
version_tuple = version.split(".")
