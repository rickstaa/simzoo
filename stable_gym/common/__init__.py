"""Contains several functions that can be used across all the Stable Gym gymnasium
environments.
"""
from stable_gym.common.disturber import Disturber  # noqa: F401
