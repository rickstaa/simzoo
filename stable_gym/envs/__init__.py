"""The Stable Gym gymnasium environments."""
