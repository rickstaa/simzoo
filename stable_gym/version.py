# Stores the package version number so that it can be accessed from other modules.
__version__ = "0.6.38"
__version_tuple__ = __version__.split(".")
