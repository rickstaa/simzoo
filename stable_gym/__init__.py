"""Module that register the Stable Gym gymnasium environments."""
import gymnasium as gym
from gymnasium.envs.registration import register

# Make module version available.
from .version import __version__, __version_tuple__

# Available environments.
# TODO: Update reward thresholds.
ENVS = {
    "Oscillator-v1": {
        "module": "stable_gym.envs.biological.oscillator.oscillator:Oscillator",
        "max_step": 400,
        "reward_threshold": 300,
    },
    "OscillatorComplicated-v1": {
        "module": "stable_gym.envs.biological.oscillator_complicated.oscillator_complicated:OscillatorComplicated",
        "max_step": 400,
        "reward_threshold": 300,
    },
    "CartPoleCost-v1": {
        "module": "stable_gym.envs.classic_control.cartpole_cost.cartpole_cost:CartPoleCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "Ex3EKF-v1": {
        "module": "stable_gym.envs.classic_control.ex3_ekf.ex3_ekf:Ex3EKF",
        "max_step": 400,
        "reward_threshold": 300,
    },
    "AntCost-v1": {
        "module": "stable_gym.envs.mujoco.ant_cost.ant_cost:AntCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "HalfCheetahCost-v1": {
        "module": "stable_gym.envs.mujoco.half_cheetah_cost.half_cheetah_cost:HalfCheetahCost",
        "max_step": 200,
        "reward_threshold": 300,
    },
    "HopperCost-v1": {
        "module": "stable_gym.envs.mujoco.hopper_cost.hopper_cost:HopperCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "SwimmerCost-v1": {
        "module": "stable_gym.envs.mujoco.swimmer_cost.swimmer_cost:SwimmerCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "Walker2dCost-v1": {
        "module": "stable_gym.envs.mujoco.walker2d_cost.walker2d_cost:Walker2dCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
}

for env, val in ENVS.items():
    if env not in gym.envs.registry:  # NOTE: Required because we use namespace packages
        register(
            id=env,
            entry_point=val["module"],
            max_episode_steps=val["max_step"],
            reward_threshold=val["reward_threshold"],
        )
