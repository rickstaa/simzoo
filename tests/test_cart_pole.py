"""Test if the CartPoleCost environment still behaves like the original CartPole
environment when the same environment parameters are used.
"""
import math
import random

import gymnasium as gym
import numpy as np
from gymnasium.logger import ERROR

import stable_gym  # noqa: F401

gym.logger.set_level(ERROR)


class TestCartPoleCostEqual:
    # Make original CartPole environment.
    env = gym.make("CartPole")
    # Make CartPoleCost environment.
    env_cost = gym.make("CartPoleCost")

    # Overwrite changed parameters.
    env_cost.unwrapped.force_mag = 10.0
    env_cost.unwrapped.theta_threshold_radians = 12 * 2 * math.pi / 360
    env_cost.unwrapped.x_threshold = 2.4
    env_cost.unwrapped.max_v = np.finfo(np.float64).max
    env_cost.unwrapped.max_x = np.finfo(np.float64).max
    env_cost.unwrapped._init_state_range = {
        "low": [-0.05, -0.05, -0.05, -0.05],
        "high": [0.05, 0.05, 0.05, 0.05],
    }

    def test_equal_reset(self):
        """Test if reset behaves the same."""
        # Perform reset and check if observations are equal.
        observation, _ = self.env.reset(seed=42)
        observation_cost, _ = self.env_cost.reset(seed=42)
        assert np.allclose(
            observation, observation_cost
        ), f"{observation} != {observation_cost}"

    def test_equal_steps(self):
        """Test if steps behave the same."""
        # Perform several steps and check if observations are equal.
        self.env.reset(seed=42), self.env_cost.reset(seed=42)
        for _ in range(10):
            discrete_action = random.randint(0, 1)
            continuous_action = (
                np.array([-10]) if discrete_action == 0 else np.array([10])
            )
            observation, _, _, _, _ = self.env.step(discrete_action)
            observation_cost, _, _, _, _ = self.env_cost.step(continuous_action)
            assert np.allclose(
                observation, observation_cost
            ), f"{observation} != {observation_cost}"

    def test_snapshot(self, snapshot):
        """Test if the 'CartPoleCost' environment is still equal to snapshot."""
        observation, info = self.env_cost.reset(seed=42)
        assert (observation == snapshot).all()
        assert info == snapshot
        self.env_cost.action_space.seed(42)
        for _ in range(5):
            action = self.env_cost.action_space.sample()
            observation, reward, terminated, truncated, info = self.env_cost.step(
                action
            )
            assert (observation == snapshot).all()
            assert reward == snapshot
            assert terminated == snapshot
            assert truncated == snapshot
            assert info == snapshot

    def test_reference_tracking_snapshot(self, snapshot):
        """Test if the 'CartPoleCost' environment with a 'reference_tracking' task is
        still equal to snapshot.
        """
        env_cost_reference_tracking = gym.make(
            "CartPoleCost",
            task_type="reference_tracking",
            exclude_reference_error_from_observation=False,
        )
        observation, info = env_cost_reference_tracking.reset(seed=42)
        assert (observation == snapshot).all()
        assert info == snapshot
        env_cost_reference_tracking.action_space.seed(42)
        for _ in range(5):
            action = env_cost_reference_tracking.action_space.sample()
            (
                observation,
                reward,
                terminated,
                truncated,
                info,
            ) = env_cost_reference_tracking.step(action)
            assert (observation == snapshot).all()
            assert reward == snapshot
            assert terminated == snapshot
            assert truncated == snapshot
            assert info == snapshot

    def test_reference_tracking_periodic_snapshot(self, snapshot):
        """Test if the 'CartPoleCost' environment with a 'reference_tracking' task and
        a 'periodic' reference is still equal to snapshot.
        """
        env_cost_reference_tracking_periodic = gym.make(
            "CartPoleCost",
            task_type="reference_tracking",
            reference_type="periodic",
            exclude_reference_error_from_observation=False,
        )
        observation, info = env_cost_reference_tracking_periodic.reset(seed=42)
        assert (observation == snapshot).all()
        assert info == snapshot
        env_cost_reference_tracking_periodic.action_space.seed(42)
        for _ in range(5):
            action = env_cost_reference_tracking_periodic.action_space.sample()
            (
                observation,
                reward,
                terminated,
                truncated,
                info,
            ) = env_cost_reference_tracking_periodic.step(action)
            assert (observation == snapshot).all()
            assert reward == snapshot
            assert terminated == snapshot
            assert truncated == snapshot
            assert info == snapshot
