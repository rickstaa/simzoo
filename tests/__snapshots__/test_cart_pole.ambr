# serializer version: 1
# name: TestCartPoleCostEqual.test_snapshot
  0.027395604855596334
# ---
# name: TestCartPoleCostEqual.test_snapshot.1
  -0.006112156024794771
# ---
# name: TestCartPoleCostEqual.test_snapshot.10
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.11
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.12
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.13
  0.03141682796673494
# ---
# name: TestCartPoleCostEqual.test_snapshot.14
  0.1589567895008271
# ---
# name: TestCartPoleCostEqual.test_snapshot.15
  0.03046562406312443
# ---
# name: TestCartPoleCostEqual.test_snapshot.16
  -0.20651144607254457
# ---
# name: TestCartPoleCostEqual.test_snapshot.17
  0.42335894938624435
# ---
# name: TestCartPoleCostEqual.test_snapshot.18
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.19
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.2
  0.03585979199113824
# ---
# name: TestCartPoleCostEqual.test_snapshot.20
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.21
  0.03459596375675149
# ---
# name: TestCartPoleCostEqual.test_snapshot.22
  0.43838373784361623
# ---
# name: TestCartPoleCostEqual.test_snapshot.23
  0.02633539514167354
# ---
# name: TestCartPoleCostEqual.test_snapshot.24
  -0.6165018621732461
# ---
# name: TestCartPoleCostEqual.test_snapshot.25
  0.31643005773596056
# ---
# name: TestCartPoleCostEqual.test_snapshot.26
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.27
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.28
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.29
  0.04336363851362381
# ---
# name: TestCartPoleCostEqual.test_snapshot.3
  0.019736802905936393
# ---
# name: TestCartPoleCostEqual.test_snapshot.30
  0.5920515317536212
# ---
# name: TestCartPoleCostEqual.test_snapshot.31
  0.014005357898208616
# ---
# name: TestCartPoleCostEqual.test_snapshot.32
  -0.8391819138660229
# ---
# name: TestCartPoleCostEqual.test_snapshot.33
  0.08976015773011449
# ---
# name: TestCartPoleCostEqual.test_snapshot.34
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.35
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.36
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.37
  0.05520466914869624
# ---
# name: TestCartPoleCostEqual.test_snapshot.38
  0.2751252402092573
# ---
# name: TestCartPoleCostEqual.test_snapshot.39
  -0.002778280379111842
# ---
# name: TestCartPoleCostEqual.test_snapshot.4
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.40
  -0.3597216590053294
# ---
# name: TestCartPoleCostEqual.test_snapshot.41
  0.004048459369276578
# ---
# name: TestCartPoleCostEqual.test_snapshot.42
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.43
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.44
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.5
  0.02727336173510044
# ---
# name: TestCartPoleCostEqual.test_snapshot.6
  0.20717331158172525
# ---
# name: TestCartPoleCostEqual.test_snapshot.7
  0.03625452804925697
# ---
# name: TestCartPoleCostEqual.test_snapshot.8
  -0.2894451993066269
# ---
# name: TestCartPoleCostEqual.test_snapshot.9
  0.5994194824080236
# ---
